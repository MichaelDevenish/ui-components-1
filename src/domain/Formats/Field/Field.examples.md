#### Basic Field

```jsx
  <Field
    label='Position Title'
  >
    Chief Executive Officer
  </Field>
```

#### Field with a Component Child

```jsx
const { Button } = require('@Domain/Buttons');
const { FontAwesomeIcon } = require('@Domain/Icons');

  <Field
    label='Position Title'
  >
    <Button
      size='small'
      type='primary-borderless'
      icon={<FontAwesomeIcon type='plus' />}
      onClick={() => alert('test')}
    >
      Add a Position Title
    </Button>
  </Field>
```
