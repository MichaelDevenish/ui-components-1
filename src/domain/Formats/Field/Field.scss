:local(.Field) {
  .label-right-component {
    padding-right: .25rem;
  }
}
