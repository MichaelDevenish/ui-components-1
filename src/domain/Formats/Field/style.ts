import styled, { css } from 'styled-components'

export interface IFieldLabelWrapperProps {
  labelRightComponent?: JSX.Element
}

export const FieldLabelWrapper = styled.div`
  ${(props: IFieldLabelWrapperProps) => props.labelRightComponent && css`
    .label-component {
      padding-right: .25rem;
    }
  `}
`
