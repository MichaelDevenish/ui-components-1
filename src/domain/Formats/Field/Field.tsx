import React, { Fragment } from 'react'
import { isString, isNumber, isNil } from 'lodash'
import { Text } from '@Domain/Typographies/Text'
import { FieldLabelWrapper } from './style'

interface IFieldProps {
  /** Label text */
  label: string

  /** Component to sit to the right of the Label */
  labelRightComponent?: JSX.Element
}

class Field extends React.PureComponent <IFieldProps> {
  public formattedChild (child: string | number | JSX.Element) : JSX.Element {
    if (isString(child) || isNumber(child)) {
      return (
        <Text>
          {child}
        </Text>
      )
    }

    return child
  }

  get formattedChildren (): JSX.Element[] {
    const {
      children
    } = this.props

    return React.Children.map(
      children,
      this.formattedChild
    )
  }

  get label (): JSX.Element {
    const {
      label,
      labelRightComponent
    } = this.props

    return (
      <FieldLabelWrapper labelRightComponent={labelRightComponent}>
        <Text
          size='small'
          color='subtle'
          isInline={!isNil(labelRightComponent)}
          className='label-component'
        >
          {label}
        </Text>
        {labelRightComponent}
      </FieldLabelWrapper>
    )
  }

  public render (): JSX.Element {
    return (
      <Fragment>
        {this.label}
        {this.formattedChildren}
      </Fragment>
    )
  }
}

export {
  IFieldProps,
  Field
}
