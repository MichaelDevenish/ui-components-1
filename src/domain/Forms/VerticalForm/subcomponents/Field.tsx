import { isString, map } from 'lodash'
import React from 'react'

import { ITooltipPopoverProps, TooltipPopover } from '../../../Popovers/TooltipPopover'
import { ErrorMessage, FieldWrapper, StyledDescription, StyledInputLabel, StyledTooltipPopover } from './style'

interface IVerticalFormFieldProps {
  /** HTML name of the input */
  inputName?: string
  /** If true, marks the input as a required field */
  isRequired?: boolean
  /** Text displayed above the input */
  label?: string
  /** Descriptive text displayed above the input */
  description?: string
  /** action text link displayed below the input */
  actionMessage?: JSX.Element
  /** Array of error messages to display */
  errorMessages?: string | string[]
  /** Message of tooltip to show */
  tooltipMessage?: JSX.Element | string
  /** any extra tooltip props */
  tooltipProps?: ITooltipPopoverProps

}

class Field extends React.PureComponent<IVerticalFormFieldProps, never> {
  private get errorMessages (): JSX.Element | Array<JSX.Element | null> | null {
    const {
      errorMessages
    } = this.props

    if (errorMessages) {
      if (isString(errorMessages)) {
        return this.errorMessageWrapper(errorMessages)
      }

      return map(errorMessages, (errorMessage: string, idx: number) => {
        return this.errorMessageWrapper(errorMessage, idx)
      })
    }

    return null
  }

  private get tooltip (): JSX.Element | null {
    const {
      tooltipMessage,
      tooltipProps
    } = this.props

    if (tooltipMessage) {
      return (
        <StyledTooltipPopover>
          <TooltipPopover {...tooltipProps}>
            {tooltipMessage}
          </TooltipPopover>
        </StyledTooltipPopover>
      )
    }

    return null
  }

  private get inputLabel (): JSX.Element | null {
    const {
      inputName,
      isRequired,
      label
    } = this.props

    if (label) {
      return (
        <StyledInputLabel
          htmlFor={inputName}
          isRequired={isRequired!}
        >
          {label}
          {this.tooltip}
        </StyledInputLabel>
      )
    }

    return null
  }

  private get description (): JSX.Element | null {
    const {
      description
    } = this.props

    if (description) {
      return (
        <StyledDescription>
          {description}
        </StyledDescription>
      )
    }

    return null
  }

  public static defaultProps = {
    isRequired: false
  }

  public render (): JSX.Element {
    const {
      children,
      actionMessage
    } = this.props

    return (
      <FieldWrapper>
        {this.inputLabel}
        {this.description}
        {children}
        {this.errorMessages}
        {actionMessage}
      </FieldWrapper>
    )
  }

  private errorMessageWrapper = (message: string, idx?: number) => {
    if (!message) {
      return null
    }

    return (
      <ErrorMessage
        key={idx || undefined}
      >
        {message}
      </ErrorMessage>
    )
  }
}

export {
  Field
}
