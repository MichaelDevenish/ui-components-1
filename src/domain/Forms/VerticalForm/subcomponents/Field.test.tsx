import { shallow } from 'enzyme'
import React from 'react'

import { Field } from './Field'

describe('<Field />', () => {
  it(`should render a vertical form field with nothing`, () => {
    const wrapper = shallow(
      <Field />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it(`should render a vertical form field with regular props`, () => {
    const wrapper = shallow(
      <Field
        inputName='testInput'
        label='This is a test input'
        isRequired
      >
        Children
      </Field>
    )

    expect(wrapper).toMatchSnapshot()
  })

  it(`should render a vertical form field with error messages`, () => {
    const wrapper = shallow(
      <Field
        label='This is a test input'
        errorMessages={[
          'Error 1',
          'Error 2'
        ]}
      >
        Children
      </Field>
    )

    expect(wrapper).toMatchSnapshot()
  })

  it(`should render a vertical form field with a single error message`, () => {
    const wrapper = shallow(
      <Field
        label='This is a test input'
        errorMessages='Just one error'
      >
        Children
      </Field>
    )

    expect(wrapper).toMatchSnapshot()
  })

  it(`should render a vertical form field with a description`, () => {
    const wrapper = shallow(
      <Field
        label='This is a test input'
        description='This is a test description'
      >
        Children
      </Field>
    )

    expect(wrapper).toMatchSnapshot()
  })

  it(`should render a vertical form field with a hint`, () => {
    const wrapper = shallow(
      <Field
        label='This is a test input'
        hintOptions={{
          hint: <div> This is a test hint in hint popover style </div>,
          label: 'This is a test hint label',
          hintWrapperProps: { width: 200 }
        }}
      >
        Children
      </Field>
    )

    expect(wrapper).toMatchSnapshot()
  })

  it(`should render a vertical form field with a action message`, () => {
    const wrapper = shallow(
      <Field
        label='This is a test input'
        actionMessage={<div>This is a test action message</div>}
      >
        Children
      </Field>
    )

    expect(wrapper).toMatchSnapshot()
  })

  it(`should render a vertical form field with a action message`, () => {
    const wrapper = shallow(
      <Field
        label='This is a test input'
        tooltipMessage='I am a test tooltip'
        tooltipProps={{width: 100}}
      >
        Children
      </Field>
    )

    expect(wrapper).toMatchSnapshot()
  })

  it(`should render a vertical form field without bottom margin`, () => {
    const wrapper = shallow(
      <Field
        inputName='testInput'
        label='This is a test input'
        showBottomMargin={false}
      >
        Children
      </Field>
    )

    expect(wrapper).toMatchSnapshot()
  })

  it(`should render a vertical form field without label`, () => {
    const wrapper = shallow(
      <Field
        inputName='testInput'
      >
        Children
      </Field>
    )

    expect(wrapper).toMatchSnapshot()
  })
})
