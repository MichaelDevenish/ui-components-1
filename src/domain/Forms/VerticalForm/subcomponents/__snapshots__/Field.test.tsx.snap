// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Field /> should render a vertical form field with a action message 1`] = `
<styled.div>
  <styled.div>
    <styled.label
      isRequired={false}
    >
      This is a test input
    </styled.label>
  </styled.div>
  Children
  <div>
    This is a test action message
  </div>
</styled.div>
`;

exports[`<Field /> should render a vertical form field with a action message 2`] = `
<styled.div>
  <styled.div>
    <styled.label
      isRequired={false}
    >
      This is a test input
      <styled.span>
        <TooltipPopover
          toggleComponent={[Function]}
          variant="neutral"
          width={100}
        >
          I am a test tooltip
        </TooltipPopover>
      </styled.span>
    </styled.label>
  </styled.div>
  Children
</styled.div>
`;

exports[`<Field /> should render a vertical form field with a description 1`] = `
<styled.div>
  <styled.div>
    <styled.label
      isRequired={false}
    >
      This is a test input
    </styled.label>
  </styled.div>
  <styled.div>
    This is a test description
  </styled.div>
  Children
</styled.div>
`;

exports[`<Field /> should render a vertical form field with a hint 1`] = `
<styled.div>
  <styled.div>
    <styled.label
      isRequired={false}
    >
      This is a test input
    </styled.label>
    <HintWrapper
      hint={
        <div>
           This is a test hint in hint popover style 
        </div>
      }
      hintType="popover"
      width={200}
    >
      <Text
        isInline={true}
        tag="span"
        type="xsmall"
      >
        This is a test hint label
      </Text>
    </HintWrapper>
  </styled.div>
  Children
</styled.div>
`;

exports[`<Field /> should render a vertical form field with a single error message 1`] = `
<styled.div>
  <styled.div>
    <styled.label
      isRequired={false}
    >
      This is a test input
    </styled.label>
  </styled.div>
  Children
  <styled.div>
    Just one error
  </styled.div>
</styled.div>
`;

exports[`<Field /> should render a vertical form field with conditional label 1`] = `
<styled.div>
  <styled.div>
    <styled.label
      conditionLabel="only test"
      htmlFor="testInput"
      isRequired={false}
    >
      This is a test input
    </styled.label>
  </styled.div>
  Children
</styled.div>
`;

exports[`<Field /> should render a vertical form field with error messages 1`] = `
<styled.div>
  <styled.div>
    <styled.label
      isRequired={false}
    >
      This is a test input
    </styled.label>
  </styled.div>
  Children
  <styled.div>
    Error 1
  </styled.div>
  <styled.div
    key="1"
  >
    Error 2
  </styled.div>
</styled.div>
`;

exports[`<Field /> should render a vertical form field with nothing 1`] = `<styled.div />`;

exports[`<Field /> should render a vertical form field with regular props 1`] = `
<styled.div>
  <styled.div>
    <styled.label
      htmlFor="testInput"
      isRequired={true}
    >
      This is a test input
    </styled.label>
  </styled.div>
  Children
</styled.div>
`;

exports[`<Field /> should render a vertical form field without bottom margin 1`] = `
<styled.div>
  <styled.div>
    <styled.label
      htmlFor="testInput"
      isRequired={false}
    >
      This is a test input
    </styled.label>
  </styled.div>
  Children
</styled.div>
`;

exports[`<Field /> should render a vertical form field without label 1`] = `
<styled.div>
  Children
</styled.div>
`;
